### DISCUSSION TITLE
Proper way to do UART re-init in the code?

### DISCUSSION THREAD

**Discussion by ghost** [micropython/micropython#11651] [ESP32] [answered]
I made simple asynchronous driver for a Neo6M GPS module -> https://github.com/divergentti/Micropython-for-ESP32/blob/main/esp32-gps/GPS_AS.py

The code seems to perform quite ok, but every now and then communication is silent or the module keeps responding "GPTXT,01,01,02,HW  UBX-G60xx  00040007 FF7FFFFFp" or GPRMC messages.

I presume this problem is related to UART somehow and I implemented self.moduleUart.init() at the end of the async loop, but that does not fix the problem.  Is there other ways to reset the UART?

HW is ESP32 with esp32-ota-20230426-v1.20.0.bin

**Comment by 2dof** [comment]
form what You said, is seems that Neo6M GPS module is resenting. What Vcc did to connected  to the module (3.3 or 5 V)?

**Reply by ghost** [reply]
3.3V from dev board. Could be issue with Dupont connectors.

**Reply by 2dof** [reply]
Some  Neo6M GPS  from China  in tx/rx module pcb line have resistors (   to let the module interface with 5V USB TTL  (lowering TTL to 3.3 V) - check what type of PCB You Got.  
If You have resistors and powering module from 3.3 then it can be  the issue.

**Reply by ghost** [reply]
VCC seems to be connected in the PCB to LB2K.  Datasheet https://www.addicore.com/NEO-6M-GPS-p/ad231.htm  "Input Supply Voltage Range: 3.3V-6V, on board voltage regulator maintains 3.3V" ... so, perhaps 3.3V output from the ESP32 dev board is drifting. I try with 5V and check if it fixes things.

**Reply by ghost** [reply]
Using 5V VCC seems to work ok.

I added RTC update from satellite (via GPRMC NMEA).

**Reply by 2dof** [reply]
My suggestions:
-  ```  def checksum(nmeaword)```   supposed  to return checkum status ( True/False) only, there is no sense to preprocessing 
  data frame: cut-off  '$' (start_code) return  and then  add again in  ``` reader()``` return.  Then if checkum ok, perform decoding from whole frame.

**Reply by ghost** [reply]
Thank you for the suggestion. I reworked and tested the code. Now simplier, less variables and simplified reader.

Do you have idea how to calculate yeardate properly? With modulo 7 I should be able to calculate weekday for RTC time setting.

OLD:
"I had to add a lot of checks to understand what is going on. I started making asynchronous code (without CRC checks etc) by following "NEO-6M GPS Module with ESP32 using MicroPython" inststructions at https://microcontrollerslab.com/neo-6m-gps-module-esp32-micropython/"

**Reply by 2dof** [reply, marked as answer]
If You mean  current date field from $GPRMC

then:
( “Doomsday Algorithm.”)
```
def weekday(y, m, d):          
        
	t = [0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4]
	y -= m < 3
	return (y + int(y/4) - int(y/100) + int(y/400) + t[m-1] + d) % 7
```
will return day no:  0  (Sunday)   ,1 (Monday) .... 
, for example :
 ``` 
wekday(23, 5, 30)   
     wekday(2023, 5, 30)  
``` 
will return value 2  (i.e Tuesday)
note that  y = 23 (2023) ,  whole year y=2023 will also work

**Comment by ghost** [comment]
Thank you for your help. The code seems to perform just fine. No crashes, no module inits.  I set latest reply as answer to original quesion.

73's 

1 ---------MCU------------- 1
   Memory free: 69872, allocated: 41296
   Heap info [(240, 0, 0, 0)], hall sensor 65, raw-temp 51.7C
2 ---------SENSOR----------- 2
   Temp: 24.6C, Rh: 32.8
   Gas: 74961.0
   Pressure: 1009.0
   Waiting GPS fix...


--- RMC debug ---
GPSTime:
System time set:  (2023, 5, 31, 8, 7, 57, 2, 151)
Data valid True:
Speed over ground: 0.573
Course over ground:
--- End of RMC ---

1 ---------MCU------------- 1
   Memory free: 73072, allocated: 38096
   Heap info [(240, 0, 0, 0)], hall sensor 66, raw-temp 51.7C
2 ---------SENSOR----------- 2
   Temp: 24.8C, Rh: 32.5
   Gas: 79148.1
   Pressure: 1009.0
   Waiting GPS fix...



1 ---------MCU------------- 1
   Memory free: 66192, allocated: 44976
   Heap info [(240, 0, 0, 0)], hall sensor 65, raw-temp 51.7C
2 ---------SENSOR----------- 2
   Temp: 24.8C, Rh: 32.3
   Gas: 82186.5
   Pressure: 1009.0
3 ---------GPS----------- 3
     Latitude: 60.358562
     Longitude: 23.100435
     Satellites: 08
     GPSTime: 08:07:57
     SystemTime: 11:08:03 and weekday: Ke


--- RMC debug ---
GPSTime: 08:07:57
System time set:  (2023, 5, 31, 8, 8, 1, 2, 151)
Data valid True:
Speed over ground: 0.144
Course over ground:
--- End of RMC ---

1 ---------MCU------------- 1
   Memory free: 69696, allocated: 41472
   Heap info [(240, 0, 0, 0)], hall sensor 61, raw-temp 51.7C
2 ---------SENSOR----------- 2
   Temp: 24.8C, Rh: 32.0
   Gas: 86687.4
   Pressure: 1009.0
3 ---------GPS----------- 3
     Latitude: 60.358570
     Longitude: 23.100435
     Satellites: 08
     GPSTime: 08:07:57
     SystemTime: 11:08:06 and weekday: Ke


--- RMC debug ---
GPSTime: 08:07:57
System time set:  (2023, 5, 31, 8, 8, 5, 2, 151)
Data valid True:
Speed over ground: 1.375
Course over ground:
--- End of RMC ---
